### PR TITLE
Add build targets for generating mocks, protobuf files, and thrift files

### PR DESCRIPTION
**Pull request by xichen2020**
cc @robskillington @mmihic 

The go generate rules are now all residing in generated/, along with the auto-generated mocks, protobuf files, and thrift files. 

I haven't quite found a good way of building proto compiler binaries as well as thrift-gen/thrift binaries because building proto compiler from source is a pain and I didn't spend time finding out which thrift-gen commit matches our existing thrift-gen/thrift binaries. For now I'm just relying on the host system to provide them. Let's talk if we should vendor these binaries tomorrow.

### CODE REVIEW

File: .ci/variables.sh [hunk 1]
```diff
@@ -1,2 +1,6 @@
+#!/bin/bash
+
 export PACKAGE='github.com/m3db/m3db'
+export VENDOR_PATH=$PACKAGE/vendor
+export LICENSE_BIN=$GOPATH/src/$VENDOR_PATH/github.com/uber/uber-licence/bin/licence
```
Comment (robskillington) [inline .ci/variables.sh]: Hm so we want to rely on Nodejs being available on your system?

File: Makefile [hunk 2]
```diff
@@ -9,6 +9,20 @@ junit_xml := junit.xml
...
+gopath_prefix := $(GOPATH)/src
+vendor_prefix := vendor
+license_dir := $(vendor_prefix)/github.com/uber/uber-licence
+license_node_modules := $(license_dir)/node_modules
+auto_gen := .ci/auto-gen.sh
+mockgen_package := github.com/golang/mock/mockgen
+mocks_output_dir := generated/mocks/mocks
+mocks_rules_dir := generated/mocks
+protoc_go_package := github.com/golang/protobuf/protoc-gen-go
+proto_output_dir := generated/proto/persistfs
```
Comment (robskillington) [inline Makefile]: Might need the ability to add more proto packages, etc.  Right now commit logs have their own proto definitions.  Or we could keep it at some high level package with a name other than `persistfs`?

File: .ci/variables.sh (hunk 1 above)
Comment (xichen2020) [inline .ci/variables.sh]: Yeah either that, or we roll out our own license fixer (e.g., some crude text prepending thing). FWIW gwr also does this.

File: Makefile (hunk 2 above)
Comment (xichen2020) [inline Makefile]: Yeah if that's the case, we can either rename the package to something like `proto`, or tweak the Makefile/bash scripts to handle multiple packages.

File: .ci/variables.sh (hunk 1 above)
Comment (robskillington) [inline .ci/variables.sh]: Kk, in future we might want to remove dependency but yeah for now I guess we just make it a requirement.

File: .ci/auto-gen.sh [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+cleanup() {
+    OLD_PATH=$PWD
+    cd $1
+    for FILE in $(ls *.go);
+    do
+        # NB(xichen): clean up vendored imports. Note that sed -i'' does not
+        # work with BSD sed shipped with OS X, whereas sed -i '' doesn't work
+        # with GNU sed, so work around it by redirecting to a temp file first
+        # and moving it back later.
+        sed "s|$VENDOR_PATH/||" $FILE > $FILE.tmp && mv $FILE.tmp $FILE
```
Comment (robskillington) [inline .ci/auto-gen.sh]: Link the mockgen issue perhaps?

**Comment by robskillington** [comment]
LGTM

**Pull request merged**